packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts

```diff
@@ -0,0 +1,149 @@
+import { Resource, TagType } from '../cloudformation/resource';
+import { Construct, IAspect } from '../core/construct';
+
+export interface TagAspectProps extends TagProperties {
+  // TODO docs
+  value?: string;
+}
+
+export interface TagProperties {
+  // TODO docs
+  propagate?: number;
+  include?: string[];
+  exclude?: string[];
+}
+
+export abstract class TagBaseAspect implements IAspect {
+
+  public readonly type: string = 'taggable';
+
+  public readonly key: string;
+  public readonly value?: string;
+
+  public propagate: number;
+  private readonly include: string[];
+  private readonly exclude: string[];
+  private readonly visitedBy: {[id: string]: boolean} = {};
+
+  private parent?: Construct;
+
+  constructor(key: string, props: TagAspectProps = {}) {
+    this.key = key;
+
+    this.propagate = props.propagate === undefined ? -1 : props.propagate;
+    this.value = props.value;
+    this.include = props.include || [];
+    this.exclude = props.exclude || [];
+  }
+
+  public visit(node: Construct): void {
+    if (this.visitedBy[node.uniqueId] === true) {
+      return;
+    }
+    this.visitedBy[node.uniqueId] = true;
+    const propagate = this.propagate > 0 ? this.propagate - 1 : this.propagate;
+    this.parent = this.parent || node;
```

Comment:
If you need this, it should also become part of the framework.

f.ex:

```ts
public visit(currentNode: Construct, applicationNode: Construct) {
```

fwiw I think it makes sense.